### PR TITLE
Add first version of collective.xkey

### PR DESCRIPTION
**Pull request by pgrunewald**

### CODE REVIEW

File: src/collective/xkey/utils.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    :type objs: object
+    :param stoponfirst: True only the first object that respond to below rules will be marked as involved.
+    :type stoponfirst: bool, optional
+    """
+    for obj in objs:
+        ids = IInvolvedIDs(obj, None)
+        if ids:
+            for single_id in ids:
+                mark_involved(request, single_id)
+            if stoponfirst:
```
Comment (veritazx) [inline src/collective/xkey/utils.py:30]: The if is one column too far to the left, it belongs inside the for.

File: src/collective/xkey/events.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        involved = get_involved_ids(request)
+        if involved:
+            request.response.setHeader("Xkey", " ".join(involved))
+
+
+@adapter(IPubAfterTraversal)
+def handle_request_after_traversal(event):
+    """handle "IPubAfterTraversal"."""
+    try:
+        # XXX: stoponfirst needs to be False?
```
Comment (veritazx) [inline src/collective/xkey/events.py]: please explain in the comment, what this is, not what it should or could be.

File: src/collective/xkey/tests/test_xkey.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+        )
+
+        browser = Browser(self.app)
+        browser.open(document.absolute_url() + "/@@special-view")
+
+        self.assertTrue("xkey" in browser.headers)
+        xkey_header = browser.headers["xkey"]
+        self.assertIn(IUUID(document), xkey_header)
+        self.assertIn(IUUID(auxiliary_document), xkey_header)
+        self.assertIn(IUUID(auxiliary_document2), xkey_header)
```
Comment (veritazx) [inline src/collective/xkey/tests/test_xkey.py:136]: Might explain, why auxiliary_document2 is supposed to be in there, when it was not explicitly marked.

File: src/collective/xkey/utils.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+def get_involved_ids(request):
+    """Get all registered involved ids from the request
+
+    :param request: request
+    :return: set of involved ids
+    :rtype: set
+    """
+    annotations = IAnnotations(request)
+    return annotations.get(KEY, set())
```
Comment (veritazx) [inline src/collective/xkey/utils.py]: Returning a (mutable) set that is not in the annotations as default might not be what you want here. Better create and set the Set here too before returning it.

File: src/collective/xkey/utils.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+
+@implementer(ITraversable)
+@implementer(IPublishTraverse)
+class MarkInvolvedView(BrowserView):
+    """Useful utility view to mark IDs in page templates.
+
+    In offers two ways for using it:
```
Comment (veritazx) [inline src/collective/xkey/utils.py:65]: It offers

File: docs/conf.py [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+project = u'collective.xkey'
+copyright = u'Paul Grunewald (pgrunewald)'
+author = u'Paul Grunewald (pgrunewald)'
+
+# The version info for the project you're documenting, acts as replacement for
+# |version| and |release|, also used in various other places throughout the
+# built documents.
+#
+# The short X.Y version.
+version = u'3.0'
```
Comment (veritazx) [inline docs/conf.py]: 3.0?

File: buildout.cfg [hunk 7]
```diff
@@ -0,0 +1,9 @@
+[buildout]
```
Comment (veritazx) [inline buildout.cfg:1]: Is this supposed to be in the Repo?

File: README.rst [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+
+    [buildout]
+
+    ...
+
+    eggs =
+        collective.xkey
+
+
+and then running ``bin/buildout``. Upon running up your instance, you may see the new header. No further setup for specific Plone sites is required.
```
Comment (veritazx) [inline README.rst]: The sentence starting with "No further..." indicates that this is usable out of the box. Does the default implementation cover all objects involved in rendering a view? What about viewlets?

File: README.rst [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+    :alt: License
+
+
+===============
+collective.xkey
+===============
+
+Use Varnish's `xkey module <https://github.com/varnish/varnish-modules>`_ for tag-based cache invalidation in Plone.
+
+This add-on is heavily inspired by and built upon on `collective.purgebyid <https://github.com/collective/collective.purgebyid>`_ (major credit goes to its contributors), but is more adapted to xkey's specific requirements.
```
Comment (veritazx) [inline README.rst]: built upon on -> built upon
, but is more -> , but it is more

File: README.rst [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+It uses tags in the same manner, but instead of banning objects with tags, it purges them with the efficient xkey implementation.
+Doing so lowers the memory consumption and improves the performance.
```
Comment (veritazx) [inline README.rst]: memory consumption and performance of the cache server, i presume? Should mention that this is only meant for the cache server. Memory consumption on Plone might rise a tiny bit.

**Review by veritazx — COMMENTED** [review]
Nice. Small fixes and docu corrections needed.

File: .github/workflows/plone-package.yml [hunk 11]
```diff
@@ -0,0 +1,50 @@
+# This workflow will install Python dependencies, run tests and lint with a variety of Python versions
+# For more information see: https://help.github.com/actions/language-and-framework-guides/using-python-with-github-actions
+
+name: Plone package
+
+on:
+    push:
+        branches: [main]
```
Comment (veritazx) [inline .github/workflows/plone-package.yml:8]: remove branches to allow any branches? tests did not run yet.

**Review by veritazx — COMMENTED** [review]
Make them run please :)

File: src/collective/xkey/utils.py (hunk 1 above)
Comment (pgrunewald) [reply src/collective/xkey/utils.py:30]: No, the intention is quite right. The stoponfirst is meant to stop after all return values for the first object. Not just the first value in general.

File: src/collective/xkey/tests/test_xkey.py (hunk 3 above)
Comment (pgrunewald) [reply src/collective/xkey/tests/test_xkey.py:136]: Take a look into `document_with_dependencies.pt`. There it is referenced:

```xml
<tal:mark-involved tal:define="auxiliary_document2 python:view.auxiliary_document2; involved auxiliary_document2/@@mark_involved" />
```

File: src/collective/xkey/utils.py (hunk 4 above)
Comment (pgrunewald) [reply src/collective/xkey/utils.py]: Done (with a slight refactor)

File: src/collective/xkey/utils.py (hunk 5 above)
Comment (pgrunewald) [reply src/collective/xkey/utils.py:65]: Fixed

File: docs/conf.py (hunk 6 above)
Comment (pgrunewald) [reply docs/conf.py]: Generated by bobtemplate, dunno why it's 3.0. Set to 1.0

File: buildout.cfg (hunk 7 above)
Comment (pgrunewald) [reply buildout.cfg:1]: Yes it is, generated by bobtemplate and makes testing easier.

File: README.rst (hunk 8 above)
Comment (pgrunewald) [reply README.rst]: The "setup" was meant no install add-on for plone site is needed. The default implementation covers alias URLs of objects. Will be clarified.

File: README.rst (hunk 9 above)
Comment (pgrunewald) [reply README.rst]: Fixed

File: README.rst (hunk 10 above)
Comment (pgrunewald) [reply README.rst]: Fixed

File: .github/workflows/plone-package.yml (hunk 11 above)
Comment (pgrunewald) [reply .github/workflows/plone-package.yml:8]: This is meant for the target branch, not the source branch. The build failed, but I'm unsure why. The tests can be run manually however.

File: src/collective/xkey/tests/test_xkey.py (hunk 3 above)
Comment (veritazx) [reply src/collective/xkey/tests/test_xkey.py:136]: jep. might be nice to have a comment here, stating that.

**Review by veritazx — APPROVED** [review]

File: src/collective/xkey/events.py (hunk 2 above)
Comment (pgrunewald) [reply src/collective/xkey/events.py]: fixed

File: src/collective/xkey/tests/test_xkey.py (hunk 3 above)
Comment (pgrunewald) [reply src/collective/xkey/tests/test_xkey.py:136]: has been added

**Pull request merged**